src/consensus/progress_map.zig

```diff
@@ -0,0 +1,1965 @@
+        if (!gop.found_existing) self.propagated_validators_stake += stake;
+        return !gop.found_existing;
+    }
+
+    /// Adds `vote_pubkey` to the propagated validator pubkey set, adding its associated `stake`
+    /// to the total propagated validator stake.
+    ///
+    /// Returns a bool indicating whether or not `vote_pubkey` was inserted into the set, ie
+    /// it returns `true` if `vote_pubkey` was inserted, and `false` if it already existed (and
+    /// thus wouldn't have been inserted).
+    ///
+    /// This assumes that the capacity of `self.propagated_validators` is enough to hold at least
+    /// one new entry, or that `vote_pubkey` is already an entry.
+    pub fn addVotePubkeyAssumeCapacity(
+        self: *PropagatedStats,
+        vote_pubkey: Pubkey,
+        stake: u64,
+    ) bool {
+        const gop = self.propagated_validators.getOrPutAssumeCapacity(vote_pubkey);
+        gop.value_ptr.* = {};
+        if (!gop.found_existing) self.propagated_validators_stake += stake;
+        return !gop.found_existing;
+    }
+
+    pub fn addNodePubkey(
+        self: *PropagatedStats,
+        allocator: std.mem.Allocator,
+        node_pubkey: Pubkey,
+        bank: stubs.Bank,
+    ) std.mem.Allocator.Error!void {
+        if (self.propagated_node_ids.contains(node_pubkey)) return;
+        const nva = bank.epochVoteAccountsForNodeId(node_pubkey) orelse return;
+        const epoch_vote_accounts = bank.epochVoteAccounts(bank.epoch()) orelse std.debug.panic(
+            "Epoch stakes for bank's own epoch must exist",
+            .{},
+        );
+        try self.addNodePubkeyInternal(
+            allocator,
+            node_pubkey,
+            nva.vote_accounts,
+            epoch_vote_accounts,
+        );
+    }
+
+    fn addNodePubkeyInternal(
+        self: *PropagatedStats,
+        allocator: std.mem.Allocator,
+        node_pubkey: Pubkey,
+        vote_account_pubkeys: []const Pubkey,
+        epoch_vote_accounts: sig.core.stake.StakeAndVoteAccountsMap,
+    ) std.mem.Allocator.Error!void {
+        try self.propagated_node_ids.put(allocator, node_pubkey, {});
+
+        try self.propagated_validators.ensureUnusedCapacity(allocator, vote_account_pubkeys.len);
+        for (vote_account_pubkeys) |vote_account_pubkey| {
+            const stake = blk: {
+                const stake, _ = epoch_vote_accounts.get(vote_account_pubkey) orelse break :blk 0;
+                break :blk stake;
+            };
+            _ = self.addVotePubkeyAssumeCapacity(vote_account_pubkey, stake);
+        }
+    }
+};
+
+pub const RetransmitInfo = struct {
+    retry_time: sig.time.Instant,
+    retry_iteration: u32,
+};
+
+pub const LockoutIntervals = struct {
+    map: HashThatShouldBeMadeBTreeMap,
+
+    /// TODO: replace this with a BTree map. In the meantime, just have to
+    /// manually keep this sorted.
+    pub const HashThatShouldBeMadeBTreeMap =
+        std.AutoArrayHashMapUnmanaged(ExpirationSlot, EntryList);
```

Comment:
What about `sig.utils.collections.SortedMap`?

Reply:
I don't want to use a managed data structure here considering that this is going to be stored in memory. That, plus there's currently no actual code that reads the field in any way that would rely on it being properly ordered, so the value gained from using the sorted map would be outweighed by the point about it being managed.